poolakey/src/main/java/ir/cafebazaar/poolakey/billing/connection/ReceiverBillingConnection.kt

```diff
@@ -0,0 +1,411 @@
+
+    override fun startConnection(context: Context, callback: ConnectionCallback): Boolean {
+        connectionCallbackReference = WeakReference(callback)
+        contextReference = WeakReference(context)
+
+        if (!Security.verifyBazaarIsInstalled(context)) {
+            return false
+        }
+
+        val bazaarVersionCode = getPackageInfo(context, BAZAAR_PACKAGE_NAME)?.let {
+            sdkAwareVersionCode(it)
+        } ?: 0L
+
+        return when {
+            canConnectWithReceiverComponent(bazaarVersionCode) -> {
+                createReceiverConnection()
+                registerBroadcast()
+                isPurchaseTypeSupported()
+                true
+            }
+            bazaarVersionCode > 0 -> {
+                callback.connectionFailed.invoke(BazaarNotSupportedException())
+                false
+            }
+            else -> {
+                false
+            }
+        }
+    }
+
+    private fun canConnectWithReceiverComponent(bazaarVersionCode: Long): Boolean {
+        return bazaarVersionCode > BAZAAR_WITH_RECEIVER_CONNECTION_VERSION
+    }
+
+    private fun createReceiverConnection() {
+        receiverCommunicator = object : BillingReceiverCommunicator {
+            override fun onNewBroadcastReceived(intent: Intent?) {
+                intent?.action?.takeIf(
+                    thisIsTrue = {
+                        isBundleSignatureValid(intent.extras)
+                    },
+                    andIfNot = {
+                        connectionCallbackReference?.get()?.connectionFailed?.invoke(
+                            PurchaseHijackedException()
+                        )
+                    }
+                )?.takeIf(
+                    thisIsTrue = {
+                        !disconnected
+                    },
+                    andIfNot = {
+                        connectionCallbackReference?.get()?.connectionFailed?.invoke(
+                            DisconnectException()
+                        )
+                    }
+                )?.let { action ->
+                    onActionReceived(action, intent.extras)
+                }
+            }
+        }
+    }
+
+    private fun onActionReceived(action: String, extras: Bundle?) {
+        when (action) {
+            ACTION_RECEIVE_BILLING_SUPPORT -> {
+                onBillingSupportActionReceived(extras)
+            }
+            ACTION_RECEIVE_CONSUME -> {
+                onConsumeActionReceived(extras)
+            }
+            ACTION_RECEIVE_PURCHASE -> {
+                onPurchaseReceived(extras)
+            }
+            ACTION_RECEIVE_QUERY_PURCHASES -> {
+                onQueryPurchaseReceived(extras)
+            }
+        }
+    }
+
+    private fun isPurchaseTypeSupported() {
+        getNewIntentForBroadcast().apply {
+            action = ACTION_BILLING_SUPPORT
+        }.run {
+            sendBroadcast(this)
+        }
```

Comment:
```kotlin
}.apply(::sendBroadcast)
```

Reply:
As we don't use the return type of this scope function, I've changed it to `run(::sendBroadcast)`
Done